compiler/damlc/tests/daml-test-files/SemanticsEvalOrder.daml

```diff
@@ -0,0 +1,362 @@
+    let _ = error "EvExpLetErr OK"
+    let _ = error "EvExpLetErr bad"
+    error "EvExpLetErr failed"
+
+-- | The following "multiple-binding let" behavior is caused by GHC.
+-- This is tracked in issue #6788. We would love to get rid of this
+-- reordering behavior, but it's very much tied with a part of GHC
+-- that we don't control. So we track this behavior here.
+ghcReordersMultiLets = scenario do
+    let _ = error "ghcReordersMultiLets failed in a good way -- close the issue & update the test"
+        _ = error "ghcReordersMultiLets OK"
+    error "ghcReordersMultiLets failed in a bad way"
+
+evExpCaseErr = scenario do
+  case error "EvExpCaseErr OK" of
+    None -> error "EvExpCaseErr failed"
+    Some _ -> error "EvExpCaseErr failed"
+
+evExpCase_1 = scenario do
+  case None of
+    None -> error "EvExpCase_1 OK"
+    Some _ -> error "EvExpCase_1 failed"
+
+evExpCase_2 = scenario do
+  case Some 1 of
+    None -> error "EvExpCase_2 failed"
+    Some _ -> error "EvExpCase_2 OK"
+
+evExpConsErr1 = scenario do
+  pure ( error "EvExpConsErr1 OK"
+      :: error "EvExpConsErr1 failed"
+      :: error "EvExpConsErr1 failed")
+
+evExpConsErr2 = scenario do
+  pure ( 10
+      :: error "EvExpConsErr2 OK"
+      :: error "EvExpConsErr2 failed")
+
+evExpBuiltinErr = scenario do
+  let _ : Int = error "EvExpBuiltinErr OK" + error "EvExpBuiltinErr failed"
+  pure ()
+
+
+data R1 = R1 { a: Int, b: Int }
+
+evExpRecConErr_1 = scenario do
+  pure R1 { a = error "EvExpRecConErr_1 OK", b = error "EvExpRecConErr_1 failed" }
+
+evExpRecConErr_2 = scenario do
+  pure R1 { b = error "EvExpRecConErr_2 failed", a = error "EvExpRecConErr_2 OK" }
+
+data R2 = R2 { d: Int, c: Int }
+  -- ^ Checking that there isn't a dependence on the field names.
+  -- There should be a dependence on the definition order according to the LF spec.
+
+evExpRecConErr_3 = scenario do
+  pure R2 { d = error "EvExpRecConErr_3 OK", c = error "EvExpRecConErr_3 failed" }
+
+evExpRecConErr_4 = scenario do
+  pure R2 { c = error "EvExpRecConErr_4 failed", d = error "EvExpRecConErr_4 OK"  }
+
+evExpRecUpdErr1 = scenario do
+  pure (error "EvExpRecUpdErr1 OK" : R1)
+    { a = error "EvExpRecUpdErr1 failed", b = error "EvExpRecUpdErr1 failed" }
+
+evExpRecUpdErr2_1 = scenario do
+  pure (R1 {a=0, b=0})
+    { a = error "EvExpRecUpdErr2_1 OK", b = error "EvExpRecUpdErr2_1 failed" }
+
+evExpRecUpdErr2_2 = scenario do
+  pure (R1 {a=0, b=0})
+    { b = error "EvExpRecUpdErr2_2 OK", a = error "EvExpRecUpdErr2_2 failed" }
+  -- ^ Note that record update depends on the order the fields appear in
+  -- code, rather than the order in which fields were defined.
+
+-- Can't test LF struct evaluation order from DAML, since we purposely avoid
+-- evaluation of struct fields during typeclass desugaring, and we don't have
+-- a way to construct LF structs directly.
+
+evExpUpPureErr = scenario do
+  let _ : Update () = pure (error "EvExpUpPureErr OK")
+  pure ()
+
+evExpUpBindErr = scenario do
```

Comment:
This feels so wrong. I'm very sad we do all this extra work to get a rather unintuitive behaviour. Nothing we can do about it now though.